File: Go/Day13.go
package main

import AH "./adventhelper"

import (
	"strconv"
	"strings"
)

type Pair struct {
	A, P int
}

func ParseLine (ss string) (pairs []Pair) {
	parts := strings.Split(ss, ",") 

	for i, s := range parts {
		if (s != "x") {
			p, _ := strconv.Atoi(s)
			pairs = append(pairs, Pair{A: i, P:p})
		}
	}

	return
}

func part1(now int, pairs []Pair) (int) {
	best := Pair{A:1000, P:0}
	for _, pair := range pairs {
		m := pair.P - (now % pair.P) 
		if m < best.A {
			best = Pair{A:m, P:pair.P}
		}
	}

	return (best.A * best.P)
}

func ChiRemThm(p1 Pair, p2 Pair) (Pair) {
	for n := 1; n <= p2.P; n++ {
		c := p1.A + n * p1.P
		if c % p2.P == p2.A % p2.P {
			return Pair{A:c, P:(p1.P*p2.P)}
		}
	}

	return Pair{A:0, P:1} // can never be hit but we can't compile without it!
}

func part2(pairs []Pair) (int) {
	out := Pair{A:0, P:1}
	for _, pair := range pairs {
		out = ChiRemThm(out, pair)
	}

	return out.P - out.A
}

func main() {
	ss, _ := AH.ReadStrFile("../input/input13.txt")
	now, _ := strconv.Atoi(ss[0])
	pairs := ParseLine(ss[1])

	AH.PrintSoln(13, part1(now, pairs), part2(pairs))

	return
}

File: Haskell/Day13.hs
import AdventHelper

import Data.List
import Data.List.Split
import Data.Function (on)

helper :: String -> [Integer]
helper s
  | s == "x"  = []
  | otherwise = [read s :: Integer]

parseBuses :: String -> [Integer]
parseBuses s = concatMap helper ss
  where ss = splitOn "," s

helper2 :: (Integer, String) -> [(Integer, Integer)]
helper2 (i, s)
  | s == "x"  = []
  | otherwise = [(i, read s :: Integer)]

parseBuses2 :: String -> [(Integer, Integer)]
parseBuses2 s = concatMap helper2 pairs
  where ss = splitOn "," s
        pairs = zip [0,1..] ss

firstBus :: Integer -> [Integer] -> (Integer, Integer)
firstBus strt bs = minPair $ zip delays bs
  where mds = map (mod strt) bs
        delays = zipWith subtract mds bs

main = do
  putStrLn "Day 10"
  f <- readFile "../input/input13.txt"
  let ss = lines f
  let now = read (head ss) :: Integer

  let buses = parseBuses (last ss)
  let part1 = firstBus now buses

  let buses2 = sortBy (compare `on` snd) $ parseBuses2 (ss!!1)
  let part2 = foldl1 chiRemThm buses2

  printSoln 1 $ fst part1 * snd part1
  printSoln 2 $ snd part2 - fst part2
